File: C++/make-the-string-great.cpp
class Solution {
public:
    string makeGood(string s) {
        bool changed = true;
        
        while (true) {
            string t = removeBad(s);
            if (t == s) {
                return t;
            }
            s = t;
        }
    }
    
    string removeBad(string s) {
        string t;
        int n = s.size();
        for (int i = 0; i < n; i++) {
            if (i < n - 1 && islower(s[i]) && isupper(s[i+1]) && s[i] == tolower(s[i+1])) {
                i++;
            } else if (i < n - 1 && isupper(s[i]) && islower(s[i+1]) && s[i] == toupper(s[i+1])) {
                i++;
            } else {
                t += s[i];
            }
        }
        return t;
    }
};

File: Go/make-the-string-great.go
func makeGood(s string) string {
	changed := true
	tmp := make([]byte, 0, len(s))
	for changed {
		changed = false
		for i := 0; i < len(s); i++ {
			if i < len(s)-1 && (s[i]-'a' == s[i+1]-'A' || s[i]-'A' == s[i+1]-'a') {
				i++
				changed = true
			} else {
				tmp = append(tmp, s[i])
			}
		}
		s = string(tmp)
		tmp = tmp[:0]
	}
	return s
}